src/sidechain.cpp

```diff
@@ -0,0 +1,466 @@
+// Copyright (c) 2017-2023 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <sidechain.h>
+
+#include <arith_uint256.h>
+#include <coins.h>
+#include <consensus/validation.h>
+#include <hash.h>
+#include <primitives/transaction.h>
+#include <streams.h>
+#include <undo.h>
+#include <util/check.h>
+
+#include <algorithm>
+#include <cstdint>
+
+void CreateDBUndoData(CTxUndo &txundo, const uint8_t type, const COutPoint& record_id, const Coin& encoded_data) {
+    Assert(record_id.n <= COutPoint::MAX_INDEX);
+    Coin& undo = txundo.vprevout.emplace_back();
+    undo = encoded_data;
+    CDataStream s(SER_NETWORK, PROTOCOL_VERSION);
+    s << record_id;
+    Assert(s.size() == 0x24);
+    auto& undo_data = undo.out.scriptPubKey;
+    undo_data.insert(undo_data.begin(), 1 + s.size(), type);
+    memcpy(&undo_data[1], &s[0], s.size());  // TODO: figure out how to jump through C++'s hoops to do this right
+}
+
+void CreateDBEntry(CCoinsViewCache& cache, CTxUndo &txundo, const int block_height, const COutPoint& record_id, const Span<const std::byte>& record_data) {
+    CScript scriptPubKey(UCharCast(record_data.begin()), UCharCast(record_data.end()));
+    cache.AddCoin(record_id, Coin(CTxOut{0, scriptPubKey}, block_height, /*fCoinbase=*/false), /*overwrite=*/false);
+
+    // Create undo data to tell DisconnectBlock to delete it
+    Coin undo;
+    CreateDBUndoData(txundo, 1, record_id, undo);
+}
+
+void DeleteDBEntry(CCoinsViewCache& inputs, CTxUndo &txundo, const COutPoint& record_id) {
+    Coin undo;
+    bool is_spent = inputs.SpendCoin(record_id, &undo);
+    assert(is_spent);
+    CreateDBUndoData(txundo, 0, record_id, undo);
+}
+
+CDataStream GetDBEntry(const CCoinsViewCache& inputs, const COutPoint& record_id) {
+    const Coin& coin = inputs.AccessCoin(record_id);
+    return CDataStream(MakeByteSpan(coin.out.scriptPubKey), SER_NETWORK, PROTOCOL_VERSION);
+}
+
+void ModifyDBEntry(CCoinsViewCache& view, CTxUndo &txundo, const int block_height, const COutPoint& record_id, const std::function<void(CDataStream&)>& modify_func) {
+    CDataStream s = GetDBEntry(view, record_id);
+    const bool new_entry = s.empty();
+    modify_func(s);
+    if (!new_entry) DeleteDBEntry(view, txundo, record_id);
+    CreateDBEntry(view, txundo, block_height, record_id, s);
+}
+
+void IncrementDBEntry(CCoinsViewCache& view, CTxUndo &txundo, const int block_height, const COutPoint& record_id, const int change) {
+    ModifyDBEntry(view, txundo, block_height, record_id, [change](CDataStream& s){
+        uint16_t counter;
+        s >> counter;
+        if (change < 0 && !counter) return;  // may be surprising if change is <-1
```

Comment:
I read the intent here as: the counter will never be < 0. If that's an important correctness property for the sidechain voting, would be nice to spell it out in a comment and/or reify it in an assert after the counter change on line 65